iothub/device/devdoc/Convention-based operations.md

```diff
@@ -0,0 +1,213 @@
+## Plug and Play convention compatible APIs
+
+#### Common
+
+```csharp
+
+public abstract class PayloadConvention {
+    protected PayloadConvention();
+    public abstract PayloadEncoder PayloadEncoder { get; }
+    public abstract PayloadSerializer PayloadSerializer { get; }
+    public virtual byte[] GetObjectBytes(object objectToSendWithConvention);
+}
+
+public abstract class PayloadEncoder {
+    protected PayloadEncoder();
+    public abstract Encoding ContentEncoding { get; }
+    public abstract byte[] EncodeStringToByteArray(string contentPayload);
+}
+
+public abstract class PayloadSerializer {
+    protected PayloadSerializer();
+    public abstract string ContentType { get; }
+    public abstract T ConvertFromObject<T>(object objectToConvert);
+    public abstract IWritablePropertyResponse CreateWritablePropertyResponse(object value, int statusCode, long version, string description = null);
+    public abstract T DeserializeToType<T>(string stringToDeserialize);
+    public abstract string SerializeToString(object objectToSerialize);
+    public abstract bool TryGetNestedObjectValue<T>(object objectToConvert, string propertyName, out T outValue);
+}
+
+public sealed class DefaultPayloadConvention : PayloadConvention {
+    public static readonly DefaultPayloadConvention Instance;
+    public DefaultPayloadConvention();
+    public override PayloadEncoder PayloadEncoder { get; }
+    public override PayloadSerializer PayloadSerializer { get; }
+}
+
+public class Utf8PayloadEncoder : PayloadEncoder {
+    public static readonly Utf8PayloadEncoder Instance;
+    public Utf8PayloadEncoder();
+    public override Encoding ContentEncoding { get; }
+    public override byte[] EncodeStringToByteArray(string contentPayload);
+}
+
+public class NewtonsoftJsonPayloadSerializer : PayloadSerializer {
+    public static readonly NewtonsoftJsonPayloadSerializer Instance;
+    public NewtonsoftJsonPayloadSerializer();
+    public override string ContentType { get; }
+    public override T ConvertFromObject<T>(object objectToConvert);
+    public override IWritablePropertyResponse CreateWritablePropertyResponse(object value, int statusCode, long version, string description = null);
+    public override T DeserializeToType<T>(string stringToDeserialize);
+    public override string SerializeToString(object objectToSerialize);
+    public override bool TryGetNestedObjectValue<T>(object objectToConvert, string propertyName, out T outValue);
+}
+
+public abstract class PayloadCollection : IEnumerable, IEnumerable<object> {
+    protected PayloadCollection();
+    public IDictionary<string, object> Collection { get; private set; }
+    public PayloadConvention Convention { get; internal set; }
```

Comment:
Is this going to stay here or move to ClientOptions?

Reply:
These payload collections (TelemetryCollection, PropertiesCollection) will have a component associated with them, so that the client knows what is the type of data these collections hold. This convention within the collection is not settable by the user, but rather is information that the sdk sets and holds internally.
The user will set the Convention via the ClientOptions.